python/ray/util/data/dataset.py

```diff
@@ -0,0 +1,359 @@
+import random
+from dataclasses import is_dataclass
+from functools import wraps
+from typing import Callable, List, Union, Iterable, Iterator
+
+import pandas as pd
+from pandas import DataFrame
+
+from ray.util.iter import (_ActorSet, _NextValueNotReady, LocalIterator, ParallelIterator,
+                           T, U)
+
+
+class MLDataset(ParallelIterator[DataFrame]):
+    """ A distributed ML dataset implemented based on ParallelIterator
+
+    All item should be a list like object or dataclass instance.
+
+    Args:
+        batch_size (int): The batch size of the current dataset. It should be
+            larger than zero, and 0 means unknown.
+    """
+
+    def __init__(self,
+                 actor_sets: List["_ActorSet"],
+                 name: str,
+                 parent_iterators: List[ParallelIterator[DataFrame]],
+                 batch_size: int):
+        super(MLDataset, self).__init__(actor_sets, name, parent_iterators)
+        self._batch_size = batch_size
+
+    @staticmethod
+    def from_parallel_it(para_it: ParallelIterator[DataFrame],
+                         batch_size: int) -> "MLDataset":
+        """
+        Create a MLDataset from an existing parallel iterator and each
+        object is a pandas.DataFrame
+        Args:
+            para_it (ParallelIterator[T]): An existing parallel iterator, and each
+                should be a list like object or dataclass instance.
+            batch_size (int): The batch size of the current dataset. It should be
+                larger than zero, and 0 means unknown.
+        Returns:
+            A MLDataset
+        """
+        return MLDataset(para_it.actor_sets, para_it.name,
+                         para_it.parent_iterators, batch_size)
+
+    def __iter__(self):
+        raise TypeError(
+            "You must use it.gather_sync() or it.gather_async() to "
+            "iterate over the results of a MLDataset.")
+
+    def __str__(self):
+        return repr(self)
+
+    def __repr__(self):
+        return f"MLDataset[{self.name}]"
+
+    def _with_transform(self, local_it_fn, name) -> "MLDataset":
+        """Helper function to create new MLDataset"""
+        para_it = super()._with_transform(local_it_fn, name)
+        return MLDataset.from_parallel_it(para_it, self._batch_size)
+
+    def transform(self, fn: Callable[[Iterable[DataFrame]], Iterable[DataFrame]]
+                  ) -> "MLDataset":
+        """
```

Comment:
pydoc style should be
```
        """Apply the fn function to the MLDataset

        Other stuff goes in new paragraph.

        Args:
```

Here and elsewhere (first line inline with quotes, then rest of comment in new paragraph).

Reply:
added